### PR TITLE
[elasticsearch] Bump version from 0.19.8 to 2.1.1

### PR DESCRIPTION
**Pull request by magicbane**
Bumped elasticsearch version and fix multiple api changes.
update README.md troubleshoot section.

Previous issue:
    ./bin/ycsb run elasticsearch -s -P workloads/workloada
    ...
    2015-12-24 02:19:55:472 4 sec: 1000 operations; 242.19 current ops/sec; [UPDATE-FAILED: Count=505, Max=15575, Min=93, Avg=286.22, 90=257, 99=2855, 99.9=9071, 99.99=15575] [READ: Count=0, Max=0, Min=9223372036854775807, Avg=�, 90=0, 99=0, 99.9=0, 99.99=0] [CLEANUP: Count=1, Max=87103, Min=87040, Avg=87072, 90=87103, 99=87103, 99.9=87103, 99.99=87103] [UPDATE: Count=0, Max=0, Min=9223372036854775807, Avg=�, 90=0, 99=0, 99.9=0, 99.99=0] [READ-FAILED: Count=495, Max=8615, Min=93, Avg=235.25, 90=243, 99=1975, 99.9=8615, 99.99=8615]
    [OVERALL], RunTime(ms), 4139.0
    [OVERALL], Throughput(ops/sec), 241.60425223483932
    [UPDATE-FAILED], Operations, 505.0
    [UPDATE-FAILED], AverageLatency(us), 286.21980198019804
    [UPDATE-FAILED], MinLatency(us), 93.0
    [UPDATE-FAILED], MaxLatency(us), 15575.0
    [UPDATE-FAILED], 95thPercentileLatency(us), 369.0
    [UPDATE-FAILED], 99thPercentileLatency(us), 2855.0
    [READ], Operations, 0.0
    [READ], AverageLatency(us), NaN
    [READ], MinLatency(us), 9.223372036854776E18
    [READ], MaxLatency(us), 0.0
    [READ], 95thPercentileLatency(us), 0.0
    [READ], 99thPercentileLatency(us), 0.0
    [READ], Return=ERROR, 495
    [CLEANUP], Operations, 1.0
    [CLEANUP], AverageLatency(us), 87072.0
    [CLEANUP], MinLatency(us), 87040.0
    [CLEANUP], MaxLatency(us), 87103.0
    [CLEANUP], 95thPercentileLatency(us), 87103.0
    [CLEANUP], 99thPercentileLatency(us), 87103.0
    [UPDATE], Operations, 0.0
    [UPDATE], AverageLatency(us), NaN
    [UPDATE], MinLatency(us), 9.223372036854776E18
    [UPDATE], MaxLatency(us), 0.0
    [UPDATE], 95thPercentileLatency(us), 0.0
    [UPDATE], 99thPercentileLatency(us), 0.0
    [UPDATE], Return=ERROR, 505
    [READ-FAILED], Operations, 495.0
    [READ-FAILED], AverageLatency(us), 235.24646464646466
    [READ-FAILED], MinLatency(us), 93.0
    [READ-FAILED], MaxLatency(us), 8615.0
    [READ-FAILED], 95thPercentileLatency(us), 412.0
    [READ-FAILED], 99thPercentileLatency(us), 1975.0
    ===
    fixed:
    [OVERALL], RunTime(ms), 38322.0
    [OVERALL], Throughput(ops/sec), 26.094671468086215
    [READ], Operations, 470.0
    [READ], AverageLatency(us), 1238.4914893617022
    [READ], MinLatency(us), 531.0
    [READ], MaxLatency(us), 10071.0
    [READ], 95thPercentileLatency(us), 2157.0
    [READ], 99thPercentileLatency(us), 4247.0
    [READ], Return=OK, 470
    [CLEANUP], Operations, 1.0
    [CLEANUP], AverageLatency(us), 240192.0
    [CLEANUP], MinLatency(us), 240128.0
    [CLEANUP], MaxLatency(us), 240255.0
    [CLEANUP], 95thPercentileLatency(us), 240255.0
    [CLEANUP], 99thPercentileLatency(us), 240255.0
    [UPDATE], Operations, 530.0
    [UPDATE], AverageLatency(us), 5101.537735849057
    [UPDATE], MinLatency(us), 2204.0
    [UPDATE], MaxLatency(us), 211199.0
    [UPDATE], 95thPercentileLatency(us), 9447.0
    [UPDATE], 99thPercentileLatency(us), 22383.0
    [UPDATE], Return=OK, 530

### CODE REVIEW

**Comment by busbey** [comment]
What's the likelihood that folks would still want to compare to the older ES versions?

File: elasticsearch/pom.xml [hunk 1]
```diff
@@ -29,10 +29,15 @@ LICENSE file.
...
     <packaging>jar</packaging>
     <properties>
-        <elasticsearch-version>0.19.8</elasticsearch-version>
+        <elasticsearch-version>2.1.1</elasticsearch-version>
     </properties>
     <dependencies>
         <dependency>
+            <groupId>net.java.dev.jna</groupId>
+            <artifactId>jna</artifactId>
+            <version>4.1.0</version>
```
Comment (busbey) [inline elasticsearch/pom.xml]: Is jna actually getting used?

File: elasticsearch/src/main/java/com/yahoo/ycsb/db/ElasticSearchClient.java [hunk 2]
```diff
@@ -90,13 +92,15 @@ public void init() throws DBException {
         .parseBoolean(props.getProperty("elasticsearch.remote", "false"));
     Boolean newdb =
         Boolean.parseBoolean(props.getProperty("elasticsearch.newdb", "false"));
-    Builder settings = settingsBuilder().put("node.local", "true")
+    Builder settings = Settings.settingsBuilder()
+        .put("node.local", "true")
```
Comment (busbey) [inline elasticsearch/src/main/java/com/yahoo/ycsb/db/ElasticSearchClient.java]: We're trying to keep datastore directions defaulted to working with a remote datastore.

Instructions in the readme for bencarking with a local node is fine,  but this change would xuse it to only work that way, right?

Comment (busbey) [inline elasticsearch/src/main/java/com/yahoo/ycsb/db/ElasticSearchClient.java]: Maybe I'm missing context due to the diff view. Let me try loading in a non-mobile client....

File: elasticsearch/src/main/java/com/yahoo/ycsb/db/ElasticSearchClient.java [hunk 3]
```diff
@@ -132,6 +144,10 @@ public void init() throws DBException {
       client = node.client();
     }
 
+    //wait for shards to be ready
+    client.admin().cluster()
+      .health(new ClusterHealthRequest("lists").waitForActiveShards(1))
+      .actionGet();
```
Comment (risdenk) [inline elasticsearch/src/main/java/com/yahoo/ycsb/db/ElasticSearchClient.java]: Is this for a local ES cluster or a remote cluster?

File: elasticsearch/src/main/java/com/yahoo/ycsb/db/ElasticSearchClient.java [hunk 4]
```diff
@@ -318,10 +333,13 @@ public Status update(String table, String key,
...
-      final RangeFilterBuilder filter = rangeFilter("_id").gte(startkey);
+      final RangeQueryBuilder filter = rangeQuery("_id").gte(startkey);
       final SearchResponse response = client.prepareSearch(indexKey)
-          .setTypes(table).setQuery(matchAllQuery()).setFilter(filter)
-          .setSize(recordcount).execute().actionGet();
+          .setTypes(table)
+          .setQuery(matchAllQuery())
+          .setSize(recordcount)
+          .execute()
+          .actionGet();
```
Comment (risdenk) [inline elasticsearch/src/main/java/com/yahoo/ycsb/db/ElasticSearchClient.java]: Looks like .setFilter() is missed here? is there a reason for that?

File: elasticsearch/pom.xml (hunk 1 above)
Comment (magicbane) [inline elasticsearch/pom.xml]: Yes, elasticsearch supports jna, though it does have a fallback in case jna isn't supported. 
I don't have a strong opinion on whether to keep it or not, so if you have a good reason to remove it I'm fine with that.

File: elasticsearch/src/main/java/com/yahoo/ycsb/db/ElasticSearchClient.java (hunk 2 above)
Comment (magicbane) [inline elasticsearch/src/main/java/com/yahoo/ycsb/db/ElasticSearchClient.java]: The default behavior for ES was set to point to local datastore. If we want to change that behavior, I'd put it in a separate commit.

File: elasticsearch/src/main/java/com/yahoo/ycsb/db/ElasticSearchClient.java (hunk 3 above)
Comment (magicbane) [inline elasticsearch/src/main/java/com/yahoo/ycsb/db/ElasticSearchClient.java]: Both. For local node it requires a short period of time for the shards to be available, otherwise the benchmark operations would start to early and cause UnavailableShardsException. In remote datastore scenario, we should assume the remote shards are available, but if they aren't, benchmarking would fail either way.

**Comment by magicbane** [comment]
the current version of ES (0.19) benchmarking is broken anyway, so I bumped up the version to 2.x.
I doubt anyone would want to benchmark 0.19.  It may be reasonable if someone would want to benchmark 1.x, but if we're fixing something, might as well bump it up to the latest stable version.

File: elasticsearch/src/main/java/com/yahoo/ycsb/db/ElasticSearchClient.java (hunk 4 above)
Comment (magicbane) [inline elasticsearch/src/main/java/com/yahoo/ycsb/db/ElasticSearchClient.java]: I goofed here. ES no longer has filter, so .setFilter() is deprecated. 
with the new API, setQuery(matchAllQuery()).setFilter(filter) should be simplified to .setQuery(rangeQuery).
Good catch.

File: elasticsearch/pom.xml (hunk 1 above)
Comment (busbey) [inline elasticsearch/pom.xml]: Interesting, so the normal client dependency will use jna if it's present but won't bring it in?

Including it is fine by me; a comment here I'm the pom about why we need to expressly include it will help avoid having it refactored out in the future.

File: elasticsearch/src/main/java/com/yahoo/ycsb/db/ElasticSearchClient.java (hunk 2 above)
Comment (busbey) [inline elasticsearch/src/main/java/com/yahoo/ycsb/db/ElasticSearchClient.java]: Sounds reasonable to me.

**Comment by busbey** [comment]
The only reason I ask about benchmarking older versions is that often we have to then maintain two bindings in order to account for differences in what a client must do to talk to the different datastore versions (e.g. take a look at the Cassandra or HBase clients). If ES doesn't have to account for this or if folks don't still compare to 1.x, then I'm all for just bumping to latest.

I believe to figure this out for Cassandra we asked on the relevant user mailing list. Is there an equivalent for ES we could check?

**Comment by magicbane** [comment]
Hmm. Considering how outdated the ES branch is, my guess is no.  I'm also not aware of any mailing list. Do you know who might know about it?

File: elasticsearch/pom.xml (hunk 1 above)
Comment (magicbane) [inline elasticsearch/pom.xml]: yes; I'll make a comment.

**Comment by busbey** [comment]
When we have a release, we announce it in the Elastic Search forum on ecosystem projects:

https://discuss.elastic.co/c/ecosystem

**Comment by magicbane** [comment]
https://discuss.elastic.co/t/ycsb-es-version-bump/38256

No one has chimed in yet.

**Comment by busbey** [comment]
4 days is enough waiting for me.

**Comment by busbey** [comment]
Thanks for the contribution!
